### PR TITLE
Add images by tag feature

### LINKED ISSUE #9: Images by Tag
#### As a user I want to view all images associated with a tag.

Acceptance criteria:
- [x] When I click on a tag, I see a filtered list of only the images that have
  the tag I clicked on.
- [x] The tag name (not its ID) is used to perform the filtering.

Discussion topics:
- [x] Where should we implement this functionality?
- [x] What happens if a user tries to manually filter by a nonexistent tag?

Dependencies:
- Image Tags

### PR DESCRIPTION
**Pull request by wpliao1989**
Closes #9

---

The following is a checklist to remind you of the steps to take when opening a
new PR. The horizontal line above and everything below can be removed from the
pull request description.

- [x] I have replaced `X` above with the associated issue number.
- [x] I have changed the issue's label to "pull request".
- [x] I have assigned the designated reviewer to this PR.
- [x] I have set the appropriate label on this PR.

If needed, please review the document on the
[Image Sharing Project](https://sites.google.com/a/appfolio.com/eng/new-hire-resources/engineering-academy/ropes/image-sharing-project)
for information on how labels should be used.

### CODE REVIEW

**Comment by wpliao1989** [comment]
Hi, please help me view this when you have a minute. @shiboying

File: Gemfile.lock [hunk 1]
```diff
@@ -240,7 +240,7 @@ DEPENDENCIES
   rails-assets-tether (>= 1.1.0)!
   rails_12factor
   sass-rails (~> 5.0)
-  selenium-webdriver
+  selenium-webdriver (= 2.53.4)
```
Comment (shiboying) [inline Gemfile.lock]: I know these changes are helpful to let you run the tests. But I think this could be better to have a separate commit. (Including the change in Gemfile)

File: app/controllers/images_controller.rb [hunk 2]
```diff
@@ -1,6 +1,11 @@
 class ImagesController < ApplicationController
   def index
-    @images = Image.order(created_at: :desc)
+    tags = params[:tag]
+    @images = if tags.present?
+                Image.tagged_with(tags).order(created_at: :desc)
+              else
+                Image.order(created_at: :desc)
+              end
```
Comment (shiboying) [inline app/controllers/images_controller.rb]: It would be great to have a flash message to warn user about not having such tag. In that case you cannot just check `tags.present?`(Since it would be also possible that the user doesn't click on any tag but just want to load the index page for all images). You can check out this `ActsAsTaggableOn::Tag.exists?(name: tag)`.

File: Gemfile.lock (hunk 1 above)
Comment (wpliao1989) [inline Gemfile.lock]: Cool, I'll squash that later.

File: app/controllers/images_controller.rb (hunk 2 above)
Comment (wpliao1989) [inline app/controllers/images_controller.rb]: Got it. I agree.

File: app/views/images/index.html.erb [hunk 3]
```diff
@@ -1,13 +1,13 @@
...
+      <div class="card">
+        <%= link_to image_path image.id do %>
           <%= image_tag(image.url, class: "center-block img-fluid image") %>
-          <%= render partial: 'image_tags', locals: { image: image } %>
-        </div>
+        <% end %>
+        <%= render partial: 'image_tags', locals: { image: image } %>
       </div>
-    <% end %>
+    </div>
```
Comment (shiboying) [inline app/views/images/index.html.erb]: I think the image show page should also show tags.

Comment (wpliao1989) [inline app/views/images/index.html.erb]: Hmmm I did that through `<%= render partial: 'image_tags', locals: { image: image } %>`

Comment (shiboying) [inline app/views/images/index.html.erb]: Oh, I see. It is just not in this PR. And another suggestions is you could check this: http://guides.rubyonrails.org/layouts_and_rendering.html 3.4.5 Rendering Collections. Then when you render partial, you could avoid using the each loop.

**Comment by wpliao1989** [comment]
Updated. I added filter section with clear filter button so a user can see what's being applied.

File: app/controllers/images_controller.rb [hunk 4]
```diff
@@ -1,6 +1,13 @@
 class ImagesController < ApplicationController
   def index
-    @images = Image.order(created_at: :desc)
+    @tags = Array.wrap(params[:tag])
```
Comment (shiboying) [inline app/controllers/images_controller.rb]: What is the reason using `Array.wrap(params[:tag])` instead of `@tags = params[:tag]`?

**Comment by shiboying** [comment]
Look good to me. Just left one more comment.

File: app/controllers/images_controller.rb (hunk 4 above)
Comment (wpliao1989) [inline app/controllers/images_controller.rb]: Yeah so I want to make sure `@tags` is an array not a string. If the query string is `?tag=cats` then `@tags == 'cats`, and if it's `?tag[]=cats&tag[]=dogs`, `@tags == ['cats', 'dogs']`

Comment (shiboying) [inline app/controllers/images_controller.rb]: Oh so you implement a muti-tags filter?

Comment (wpliao1989) [inline app/controllers/images_controller.rb]: There's no UI for multi tags but one can do that with query string.

Comment (shiboying) [inline app/controllers/images_controller.rb]: Got it. Thanks. I think the pr is good to go.

Comment (wpliao1989) [inline app/controllers/images_controller.rb]: Cool! Thanks for your feedback! I'm going to squash commits then merge it.
